cluster/network.go

```diff
@@ -209,3 +231,236 @@ func (c *Cluster) getNetworkPluginManifest(pluginConfig map[string]string) (stri
 		return "", fmt.Errorf("[network] Unsupported network plugin: %s", c.Network.Plugin)
 	}
 }
+
+func (c *Cluster) CheckClusterPorts(ctx context.Context) error {
+	if err := c.deployTCPPortListeners(ctx); err != nil {
+		return err
+	}
+	if err := c.runServicePortChecks(ctx); err != nil {
+		return err
+	}
+	if err := c.checkKubeAPIPort(ctx); err != nil {
+		return err
+	}
+	return c.removeTCPPortListeners(ctx)
+}
+
+func (c *Cluster) checkKubeAPIPort(ctx context.Context) error {
+	log.Infof(ctx, "[network] Checking KubeAPI port Control Plane hosts")
+	for _, host := range c.ControlPlaneHosts {
+		logrus.Debugf("[network] Checking KubeAPI port [%s] on host: %s", KubeAPIPort, host.Address)
+		address := fmt.Sprintf("%s:%s", host.Address, KubeAPIPort)
+		conn, err := net.Dial("tcp", address)
+		if err != nil {
+			return fmt.Errorf("[network] Can't access KubeAPI port [%s] on Control Plane host: %s", KubeAPIPort, host.Address)
+		}
+		conn.Close()
+	}
+	return nil
+}
+
+func (c *Cluster) deployTCPPortListeners(ctx context.Context) error {
+	log.Infof(ctx, "[network] Deploying port listener containers")
+	var errgrp errgroup.Group
+
+	portList := []string{
+		KubeAPIPort,
+		EtcdPort1,
+		EtcdPort2,
+		ScedulerPort,
+		ControllerPort,
+		KubeletPort,
+		KubeProxyPort,
+	}
+	portBindingList := []nat.PortBinding{}
+	for _, portNumber := range portList {
+		rawPort := fmt.Sprintf("0.0.0.0:%s:1337/tcp", portNumber)
+		portMapping, _ := nat.ParsePortSpec(rawPort)
+		portBindingList = append(portBindingList, portMapping[0].Binding)
+	}
+
+	imageCfg := &container.Config{
+		Image: c.SystemImages[AplineImage],
+		Cmd: []string{
+			"nc",
+			"-kl",
+			"-p",
+			"1337",
+			"-e",
+			"echo",
+		},
+		ExposedPorts: nat.PortSet{
+			"1337/tcp": {},
+		},
+	}
+
+	hostCfg := &container.HostConfig{
+		PortBindings: nat.PortMap{
+			"1337/tcp": portBindingList,
+		},
+	}
+	hostList := []*hosts.Host{}
+	hostList = append(hostList, c.EtcdHosts...)
+	hostList = append(hostList, c.ControlPlaneHosts...)
+	hostList = append(hostList, c.WorkerHosts...)
+	// little trick to get a unique host list
+	uniqHosts := make(map[*hosts.Host]bool)
```

Comment:
why pointer as a key?

Reply:
Because host lists are list of host pointers, not host objects.